### ISSUE TITLE
foreach optimizer pattern for much faster @compile with large number of parameter leafs

### ISSUE THREAD

**Issue by csvance** [EnzymeAD/Reactant.jl#2929]
Currently if you try to `@compile` the optimizer update step for a model with many parameter leafs, it takes a massive amount of time because Reactant/XLA has to handle the exact same element wise update across every leaf individually. With models like ResNet50 and Adam optimizer variants, I've observed compile times close to an hour just for the optimizer step. 

The solution is to flatten all parameters and handle the update in a single step which is how PyTorch handles it with its foreach flag. There isn't much documentation about this as far as I can tell, so it might be worth some sort of abstraction/utility functions either here or in downstream packages to automatically handle this seamlessly. If this sort of thing just worked without hour long compile times, it would be much easier to pickup Reactant.jl and be immediately productive.

**Comment by wsmoses** [comment]
yeah if you can share a reproducer for this, we should fix this

**Comment by csvance** [comment]
I can easily reproduce ~150 leafs adding two minutes, but I'm trying to isolate what was causing an entire hour now. The two minutes could probably be addresses as a documentation gap in other packages, its just more noticeable in Reactant because of the longer compile times. The hour is most likely something strange I'm doing or a bug.

**Comment by avik-pal** [comment]
(Haven't validated the hypothesis) It is possible that our tracing over https://github.com/FluxML/Functors.jl is causing the pathology.

**Comment by csvance** [comment]
@wsmoses it seems like this could be folded into a dedicated documentation page for PyTorch users. Something like "Getting Started for PyTorch Users". It could go over concepts like DP/DDP training and how that maps into Reactant, common pitfalls working with optimisers, etc. Since so many people use PyTorch, I would imagine it would be helpful to people wanting to give Reactant a try.

**Comment by wsmoses** [comment]
sure, go for it -- though also we should fix this regardless to not be slow

**Comment by csvance** [comment]
@wsmoses I have been developing an ML framework ontop of Reactant which automatically handles this and structures problems in a way which avoids many common pitfalls and reduces boilerplate:

- Flattens parameter tree before applying optimizer update rule to speed up compile; parameter groups, per parameter weight decay, and even multiple optimizers are supported out of the box. Plays nicely with almost everything in Optimisers.jl.
- Convention that explicitly hides things from Reactant/Enzyme which might otherwise be automatically used as a const and cause massive compile time increase. This typically happens when the user has a lot of data associated with producing the actual training examples like dataframes, indexes, etc and that gets passed into the `@compile` even though the traced function doesn't use them.
- Doesn't rely on GC to free batch device buffers; they are freed immediately after each batch. You don't go OoM when there isn't host GC pressure. Uses the same method as ReactantServer for this, it might actually be worth providing a formal interface in Reactant for this.
- Avoid the recompile pitfall for things like learning rate schedules, teacher forcing, etc by explicitly marking certain experiment parameters as "Tunable{T}" which can have schedules / update rules defined on the host side. Values that the user wants to sweep over should all be defined as Tunable{T} instead of a constant in Julia.
- Metrics on either host or device for both train and evaluation. Host metrics are more flexible and let you not pay a recompile every time you change them or add new ones at the expense of paying more host->device copies. Training defaults to device, validation/test to host. 
- REPL driven interface where pretty much anything can be overridden as a kwarg, including tunable parameters. Single statements are provided to train, validate, test, predict, etc.
- Supports gradient accumulation if the desired batch size cannot fit across the available devices (why we have two separate programs for training, in the future we could just fuse everything when not doing gradient accumulation which I already did in v1)
- Provides simple but extensible abstractions for loggers

I just did a big design effort for a V2 which basically takes all of the things that were specific to my workflows and replaces them with something more suitable for general usage. Is this a project that could make sense in EnzymeAD? I feel like there is a place for a Reactant centric ML framework in Julia that builds ontop of Reactant/Lux at a higher level and structures things in a way that is much faster to get an experiment off the ground.

Here is a gist produced by Claude showing the framework conventions: https://gist.github.com/csvance/ade82bf748e697ff13d458a90f7f9fcc

**Comment by wsmoses** [comment]
Most definitely, and let us know how we can help [that also concurrently said we should try to fix any pitfalls here too that are reasonable to do so]
